SWLOR.Game.Server/Feature/AIDefinition/GenericAIDefinition.cs

```diff
@@ -72,6 +171,402 @@ static float CalculateAverageHP(uint creature)
+            {
+                return (FeatType.ForceDrain3, target);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForceDrain4, () => activeConcentration == FeatType.Invalid))
+            {
+                return (FeatType.ForceDrain4, target);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForceDrain5, () => activeConcentration == FeatType.Invalid))
+            {
+                return (FeatType.ForceDrain5, target);
+            }
+
+            // Force Push
+            if (CheckIfCanUseFeat(self, target, FeatType.ForcePush1))
+            {
+                return (FeatType.ForcePush1, target);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForcePush2))
+            {
+                return (FeatType.ForcePush2, target);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForcePush3))
+            {
+                return (FeatType.ForcePush3, target);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForcePush4))
+            {
+                return (FeatType.ForcePush4, target);
+            }
+
+            // Force Inspiration
+            if (CheckIfCanUseFeat(self, self, FeatType.ForceInspiration1))
+            {
+                return (FeatType.ForceInspiration1, self);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForceInspiration2))
+            {
+                return (FeatType.ForceInspiration2, self);
+            }
+            if (CheckIfCanUseFeat(self, target, FeatType.ForceInspiration3))
+            {
+                return (FeatType.ForceInspiration3, self);
+            }
+
+            // Shocking Shout
+            if (CheckIfCanUseFeat(self, self, FeatType.ShockingShout))
+            {
+                return (FeatType.ShockingShout, self);
+            }
+
+            // Soldier's Precision
+            if (CheckIfCanUseFeat(self, self, FeatType.SoldiersPrecision))
+            {
+                return (FeatType.SoldiersPrecision, self);
+            }
+
+            // Charge
+            if (CheckIfCanUseFeat(self, self, FeatType.Charge))
+            {
+                return (FeatType.Charge, self);
+            }
+
+            // Soldier's Strike
+            if (CheckIfCanUseFeat(self, self, FeatType.SoldiersStrike))
+            {
+                return (FeatType.SoldiersStrike, self);
+            }
+
+            // Rejuvination
+            if (CheckIfCanUseFeat(self, self, FeatType.Rejuvenation))
+            {
+                return (FeatType.Rejuvenation, self);
+            }
+
+            // Frenzied Shout
+            if (CheckIfCanUseFeat(self, self, FeatType.FrenziedShout))
+            {
+                return (FeatType.FrenziedShout, self);
+            }
+
+            // Soldier's Speed
+            if (CheckIfCanUseFeat(self, self, FeatType.SoldiersSpeed))
```

Comment:
I'm pretty sure the AI is going to enable these auras initially and then as soon as the recast timer is up, they'll use it again which will disable it. You probably need to add a condition to ensure this doesn't happen.

Reply:
To be honest I haven't the faintest clue how to do that :(

Reply:
This is going to require some refactoring. The Auras are handled from a base class which won't be accessible by the AI service. Once that's moved out, I think what we'd do here is at the start of the AI cycle, determine how many auras are active and how many they support (using existing rules). Once we have that in place it'd probably be something like:

if(CheckIfCanUseFeat(self, self, FeatType.SoldiersSpeed, () => aurasActive < maxAuras && !activeAuras.Contains(FeatType.SoldiersSpeed))

(total pseudocode here)

Because this is pretty involved I would recommend leaving out the auras from here for the time being.